### PR TITLE
Recipe – Serve a File from the Back-End

### PR DESCRIPTION
**Pull request by 4c3x5u**
@isaacabraham @theimowski Hi. Could one of you guys review this when free please? :)

### CODE REVIEW

**Comment by Zaid-Ajaj** [comment]
Hi Alican, can you please update the downloading part with when using Fable.Remoting to use the latest bits available in the Fable.Remoting.Client package which now includes a SaveFileAs extension function for byte array as explained in the docs https://zaid-ajaj.github.io/Fable.Remoting/src/upload-and-download.html

The template will need to update Fable.Remoting.Client to V6.12.0

**Comment by functionalprogrammer** [comment]
@Zaid-Ajaj Ah. This is great. I was wondering whether there wasn't a simpler way to do this, to be honest. I'll do the changes now. Thank you!

**Comment by Zaid-Ajaj** [comment]
Well, I added the feature over the weekend 😉 (I also sent you a message on slack to tell about it) 

Can you also reference a link to the document "File Upload And Download" from the recipe to Fable.Remoting docs?

**Comment by functionalprogrammer** [comment]
Ahaha, that's my bad then. Sure. 😄

**Comment by functionalprogrammer** [comment]
Just implemented the changes suggested by @Zaid-Ajaj.

File: docs/recipes/client-server/serve-a-file-from-the-backend.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+#### 4. Using the download funciton
+
+Since the `downloadFile` function is asynchronous, we can't just call it anywhere in our view. The way we're going to deal with this is to create a `Msg` case and handle it in our `update` funciton.
+
+##### a. Add a couple of new cases to the Msg type
+
+```fsharp
+type Msg =
+    //...other cases
+    DownloadFile
```
Comment (theimowski) [inline docs/recipes/client-server/serve-a-file-from-the-backend.md:129]: Missing pipe `|` symbols before DU cases' definitions

File: docs/recipes/client-server/serve-a-file-from-the-backend.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+Although not perfect, the best solution for handling the file download is creating an invisible download link, clicking it, and then removing it completely. The following block of code is all we need for this. Add it to the *Index.fs* file, somewhere above the `view` function.
+
+```fsharp
+open Fable.Core.JsInterop
+
+let downloadFile () =
+    let anchor = Browser.Dom.document.createElement "a"
+    anchor?style <- "display: none"
+    anchor?href <- "/api/file"
```
Comment (theimowski) [inline docs/recipes/client-server/serve-a-file-from-the-backend.md:49]: use `Route.file` from shared

**Comment by theimowski** [comment]
In both examples we demonstrate reading whole file into memory.
I think we should show how to do streaming at least for the minimal template (as an "more advanced" usage):

```fsharp
        get Route.file (fun next ctx ->
            task {
                let filePath = "..."
                return! ctx.WriteFileStreamAsync false filePath None None
            })
```

And then emphasise that for default sample whole file would be in memory - BTW @Zaid-Ajaj wondering if one can achieve similar streaming using Fable.Remoting?

**Comment by Zaid-Ajaj** [comment]
> wondering if one can achieve similar streaming using Fable.Remoting?

@theimowski I am afraid that is not possible at the moment, we can't use `Stream` as a shared type between client and server. I think it should be possible to do streaming but I'll have to think about the shared API first 🤔

**Pull request merged**